server/src/main/java/com/thoughtworks/go/server/newsecurity/x509/CachingSubjectDnX509PrincipalExtractor.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2018 ThoughtWorks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.thoughtworks.go.server.newsecurity.x509;
+
+import net.sf.ehcache.*;
+import net.sf.ehcache.config.CacheConfiguration;
+import net.sf.ehcache.store.MemoryStoreEvictionPolicy;
+import org.springframework.dao.DataRetrievalFailureException;
+import org.springframework.security.ui.preauth.x509.SubjectDnX509PrincipalExtractor;
+import org.springframework.security.ui.preauth.x509.X509PrincipalExtractor;
+
+import java.security.cert.X509Certificate;
+
+public class CachingSubjectDnX509PrincipalExtractor implements X509PrincipalExtractor {
+
+    private static final String CACHE_NAME = "agentCertificateCache";
+    private final SubjectDnX509PrincipalExtractor delegate = new SubjectDnX509PrincipalExtractor();
+
+    private final Ehcache cache;
+
+    public CachingSubjectDnX509PrincipalExtractor() {
+        this(createCacheIfRequired());
+    }
+
+    private static Ehcache createCacheIfRequired() {
```

Comment:
The cache has been internalized to simplify things and reduce yet another spring component.